tests/test_salesforce_incremental_table_reset.py

```diff
@@ -0,0 +1,145 @@
+        first_sync_record_count = self.run_and_verify_sync(conn_id)
+        first_sync_records = runner.get_records_from_target_output()
+
+        first_sync_bookmarks = menagerie.get_state(conn_id)
+
+        # UPDATE STATE for Table Reset
+        new_states = {'bookmarks': dict()}
+        for stream, new_state in self.calculated_states_by_stream(first_sync_bookmarks).items():
+            replication_key = list(replication_keys[stream])[0]
+            # Remove stream User to simulate table reset
+            if stream != 'User':
+                new_states['bookmarks'][stream] = {replication_key: new_state}
+        menagerie.set_state(conn_id, new_states)
+
+        # SYNC 2
+        second_sync_record_count = self.run_and_verify_sync(conn_id)
+        second_sync_records = runner.get_records_from_target_output()
+        second_sync_bookmarks = menagerie.get_state(conn_id)
+
+        # Test by stream
+        for stream in expected_streams:
+            with self.subTest(stream=stream):
+                # record counts
+                first_sync_count = first_sync_record_count.get(stream, 0)
+                second_sync_count = second_sync_record_count.get(stream, 0)
+
+                # data from record messages
+                first_sync_messages = [record.get('data') for record in
+                                       first_sync_records.get(stream).get('messages')
+                                       if record.get('action') == 'upsert']
+                second_sync_messages = [record.get('data') for record in
+                                        second_sync_records.get(stream).get('messages')
+                                        if record.get('action') == 'upsert']
+
+                # replication key for comparing data
+                self.assertEqual(1, len(list(replication_keys[stream])),
+                                 msg="Compound primary keys require a change to test expectations")
+                replication_key = list(replication_keys[stream])[0]
+
+                # bookmarked states (top level objects)
+                first_bookmark_key_value = first_sync_bookmarks.get('bookmarks').get(stream)
+                second_bookmark_key_value = second_sync_bookmarks.get('bookmarks').get(stream)
+
+                # bookmarked states (actual values)
+                first_bookmark_value = first_bookmark_key_value.get(replication_key)
+                second_bookmark_value = second_bookmark_key_value.get(replication_key)
+
+                # bookmarked values as utc for comparing against records
+                first_bookmark_value_utc = self.convert_state_to_utc(first_bookmark_value)
+                second_bookmark_value_utc = self.convert_state_to_utc(second_bookmark_value)
+
+                # Verify the second sync bookmark is Equal to the first sync bookmark
+                self.assertEqual(second_bookmark_value, first_bookmark_value) # assumes no changes to data during test
```

Comment:
I'd suggest either using our filter pattern here or asserting greaterEqual to allow for data coming in between syncs as we have seen this be a problem in the past.

Reply:
Changed the assertion to greaterEqual

Reply:
Added filtering to take care of record inserts between 2 syncs